core/domain/translation_domain.py

```diff
@@ -18,9 +18,285 @@
+
+
+class TranslatedContent:
+    """Class representing a translation of translatable content. For example,
+    if translatable content 'A' is translated into 'B' in a language other than
+    English, then 'B' is a TranslatedContent instance that represents this
+    class.
+    A (TranslatableContent) -----(translation)-----> B (TranslatedContent).
+
+    Args:
+        content_value: ContentValueType. Represents translation of translatable
+            content.
+        needs_update: bool. Whether any translation needs an update or not.
+    """
+
+    def __init__(
+        self,
+        content_value: feconf.ContentValueType,
+        needs_update: bool
+    ) -> None:
+        self.content_value = content_value
+        self.needs_update = needs_update
+
+    @classmethod
+    def from_dict(
+        cls,
+        translated_content_dict: feconf.TranslatedContentDict
+    ) -> TranslatedContent:
+        """Returns a TranslatedContent object from its dict representation.
+
+        Args:
+            translated_content_dict: TranslatedContentDict. Dict representation
+                of TranslatedContent object.
+
+        Returns:
+            TranslatedContent. An instance of TranslatedContent class.
+        """
+        return cls(
+            translated_content_dict['content_value'],
+            translated_content_dict['needs_update'])
+
+    def to_dict(self) -> feconf.TranslatedContentDict:
+        """Returns the dict representation of TranslatedContent object.
+
+        Returns:
+            TranslatedContentDict. A dict, mapping content_value and
+            needs_update of a TranslatableContent instance to
+            corresponding keys 'content_value' and 'needs_update'.
+        """
+        return {
+            'content_value': self.content_value,
+            'needs_update': self.needs_update
+        }
+
+
+class TranslatableContentsCollection:
+    """A class to collect all TranslatableContents from a translatable object
+    and map them with their corresponding content-ids.
+    """
+
+    content_id_to_translatable_content: Dict[str, TranslatableContent]
+
+    def __init__(self) -> None:
+        """Constructs a TranslatableContentsCollection object."""
+        self.content_id_to_translatable_content = {}
+
+    def add_translatable_field(
+        self,
+        field_type: TranslatableContentFormat,
+        content_id: str,
+        content_value: feconf.ContentValueType
+    ) -> None:
+        """Adds translatable field parameter to
+        'content_id_to_translatable_content' dict.
+
+        Args:
+            field_type: TranslatableContentFormat. The type of the
+                corresponding translatable content.
+            content_id: str. The id of the corresponding translatable content.
+            content_value: ContentValueType. Value of the content which
+                is translatable.
+        """
+        # The duplicate check is avoided here as object validation check ensures
```

Comment:
Suggest: "We aren't checking for duplicate values here, since the validate() check for each object will ensure that all the content_ids are unique." 

(I.e. being more explicit about what the "object validation check" is, and providing more context for the "duplicate check".)

Alternatively (and probably better for the time being, pragmatically?) -- don't make this assumption, and instead raise an exception if a content ID that already exists as a key in the dict is added. Ditto below.

Reply:
Added if block to check whether a content-id is already added in the dict or not.